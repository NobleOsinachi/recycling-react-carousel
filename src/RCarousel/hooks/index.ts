export * from './useOnResize';
export * from './usePrevious';
export * from './useWindowWidth';
export * from './useForceUpdate';
export * from './useShouldUpdate';
export * from './useRCalc';
