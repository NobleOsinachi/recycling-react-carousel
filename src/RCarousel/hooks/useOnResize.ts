import { useEffect, useState, useRef, RefObject, useCallback } from 'react';

type Size = {
  width: number;
  height: number;
};

export function useOnResize(staticRefs: RefObject<HTMLElement | null>[] = []) {
  const [sizes, setSizes] = useState<Size[]>([]);
  const refs = useRef(staticRefs);
  const observeRef = useRef<ResizeObserver>(
    new ResizeObserver(() => {
      const nodes = refs.current
        .filter(({ current }) => current)
        .map(({ current }) => current) as HTMLElement[];

      setSizes(
        nodes.map(({ offsetWidth: width, offsetHeight: height }) => ({
          width,
          height,
        }))
      );
    })
  );

  const update = useCallback(() => {
    const ro = observeRef.current;
    ro.disconnect();
    refs.current.forEach(({ current: node }) => node && ro.observe(node));
  }, []);

  const add = useCallback((ref: RefObject<HTMLElement>) => {
    const ro = observeRef.current;
    ro.observe(ref.current as Element);
    refs.current.push(ref);
  }, []);

  const remove = useCallback((ref: RefObject<HTMLElement>) => {
    const ro = observeRef.current;
    const i = refs.current.indexOf(ref);
    if (~i) {
      ro.unobserve(refs.current[i].current as Element);
      refs.current.splice(i, 1);
    }
  }, []);

  useEffect(() => {
    const ro = observeRef.current;

    update();

    return () => ro.disconnect();
  }, [update]);

  return { add, remove, update, sizes };
}
