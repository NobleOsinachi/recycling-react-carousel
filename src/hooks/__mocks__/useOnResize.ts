import React from 'react';

export function useOnResize(containerRef: React.RefObject<HTMLDivElement>) {
  const [size, setSize] = React.useState([0, 0]);
  React.useEffect(() => {
    if (!containerRef.current) {
      return;
    }
    setSize([window.innerWidth, window.innerHeight]);
  }, [containerRef]);
  return size;
}
