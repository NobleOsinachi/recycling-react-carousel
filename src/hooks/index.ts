export * from './useCursor';
export * from './useOnResize';
export * from './usePrevious';
export * from './useWindowWidth';
export * from './useForceUpdate';
export * from './useShouldUpdate';
