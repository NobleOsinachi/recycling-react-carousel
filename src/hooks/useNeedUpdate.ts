import { usePrevious } from './usePrevious';

export function useNeedUpdate(...deps: any[]) {
  const curToken = JSON.stringify(deps);
  const prevToken = usePrevious(curToken);
  return curToken !== prevToken;
}
