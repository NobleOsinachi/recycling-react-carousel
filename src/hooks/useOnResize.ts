import { useEffect, useState, useRef, RefObject } from 'react';

export function useOnResize(containerRef: RefObject<HTMLDivElement>) {
  const [size, setSize] = useState({ width: 0, height: 0 });
  const observeRef = useRef<ResizeObserver>();

  useEffect(() => {
    if (!containerRef.current) {
      return;
    }

    observeRef.current = new ResizeObserver(() => {
      if (containerRef.current) {
        const {
          offsetWidth: width,
          offsetHeight: height,
        } = containerRef.current;
        setSize({ width, height });
      }
    });

    observeRef.current.observe(containerRef.current);

    return () => observeRef.current?.disconnect();
  }, [containerRef]);

  return size;
}
