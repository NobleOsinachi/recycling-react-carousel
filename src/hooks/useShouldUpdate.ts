import { usePrevious } from './usePrevious';
import isEqual from 'lodash/isEqual';

export function useShouldUpdate(...curDeps: any[]) {
  const prevDeps = usePrevious(curDeps);
  return !isEqual(curDeps, prevDeps);
}
