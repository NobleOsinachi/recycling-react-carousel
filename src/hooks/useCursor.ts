import { useState, useCallback, useMemo } from 'react';
import { clampCursor, getLocalIndex } from '../RCarousel/helpers';

interface CursorProps {
  init?: number;
  step?: number;
}

export function useCursor({ init = 0, step = 1 }: CursorProps = {}) {
  const [[from, to, slidesCount], setRange] = useState<number[]>([0, 0, 0]);
  const [globalCursor, set] = useState(init);
  const getLocalCursor = (globalCursor: number, slidesCount: number) =>
    getLocalIndex(globalCursor, slidesCount);
  const localCursor = getLocalCursor(globalCursor, slidesCount);

  const clampAndSet = useCallback(
    (value: number | ((arg: number) => number)) => {
      set((prev: number) =>
        clampCursor(typeof value === 'function' ? value(prev) : value, from, to)
      );
    },
    [from, to]
  );

  const goTo = useCallback(
    (value: number) =>
      set(
        (prevGlobalCursor: number) =>
          clampCursor(value, 0, slidesCount - 1) -
          getLocalCursor(prevGlobalCursor, slidesCount) +
          prevGlobalCursor
      ),
    [slidesCount]
  );

  const move = useCallback(
    (step: number) => () =>
      clampAndSet((globalCursor: number) => globalCursor + step),
    [clampAndSet]
  );

  const onRangeChange = useCallback(
    (...range: number[]) => setRange(range),
    []
  );

  const next = useMemo(() => move(1 * step), [move, step]);
  const back = useMemo(() => move(-1 * step), [move, step]);

  const isMax = !(globalCursor + step <= to);
  const isMin = !(globalCursor - step >= from);

  return {
    cursor: localCursor,
    clampAndSet,
    goTo,
    move,
    next,
    back,
    firstIndex: from,
    lastIndex: to,
    length: slidesCount,
    isMax,
    isMin,
    props: {
      cursor: globalCursor,
      onCursorChange: set,
      onRangeChange,
    },
    nextBtnProps: { onClick: next, disabled: isMax },
    backBtnProps: { onClick: back, disabled: isMin },
  };
}
