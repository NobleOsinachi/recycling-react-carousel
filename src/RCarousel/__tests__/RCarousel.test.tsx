import React from 'react';
import { buildTestCarousel, getTestSlides, swipe, render } from './tools';
import { useOnResize as mockedUseOnResize } from '../../hooks/useOnResize';

import './tools/patchCreateEvent';

jest.mock('../../hooks/useOnResize');
// jest.mock('../helpers/animTo');

const useOnResize = mockedUseOnResize as jest.Mock;
const getUseOnResizeMock = (sizes: any = []) => () => ({
  sizes,
  add: jest.fn(),
  remove: jest.fn(),
});

const DEFAULT_WINDOW_WIDTH = 1000;
const DEFAULT_WINDOW_HEIGHT = 200;

const resize = (
  width = DEFAULT_WINDOW_WIDTH,
  height = DEFAULT_WINDOW_HEIGHT
) => {
  window.resizeTo(width, height);
  useOnResize.mockImplementation(getUseOnResizeMock([{ width, height }]));
};

beforeAll(() => {
  window.resizeTo = function resizeTo(width, height) {
    Object.assign(this, {
      innerWidth: width,
      innerHeight: height,
      outerWidth: width,
      outerHeight: height,
    }).dispatchEvent(new this.Event('resize'));
  };
});

beforeEach(() => {
  resize();
});

test('check defaultCursor prop', () => {
  const { Component } = buildTestCarousel();
  const { asFragment } = render(<Component defaultCursor={2} />);
  expect(asFragment()).toMatchSnapshot();
});

test('check carousel without slides', () => {
  const { Component } = buildTestCarousel({ slidesCount: 0 });
  const { asFragment } = render(<Component />);
  expect(asFragment()).toMatchSnapshot();
});

test('check carousel with only one slide', () => {
  const { Component } = buildTestCarousel({ slidesCount: 1 });
  const { asFragment } = render(<Component />);
  expect(asFragment()).toMatchSnapshot();
});

describe('check slides switching using cursor prop', () => {
  const { Component } = buildTestCarousel();

  test.each([
    [0, 1],
    [1, 2],
    [2, 1],
    [1, 0],
  ])('move from cursor %i to cursor %i', (from, to) => {
    const { rerender, asFragment } = render(<Component cursor={from} />);
    const curFragment = asFragment();
    rerender(<Component cursor={to} />);
    expect(curFragment).toMatchDiffSnapshot(asFragment());
  });
});

describe('check slides switching on swiping (X axis)', () => {
  const changeHandler = jest.fn();
  const { Component } = buildTestCarousel({
    baseProps: {
      onCursorChange: changeHandler,
    },
  });

  test.each([
    [DEFAULT_WINDOW_WIDTH, 0, 0, 1],
    [DEFAULT_WINDOW_WIDTH, 0, 1, 2],
    [0, DEFAULT_WINDOW_WIDTH, 2, 1],
    [0, DEFAULT_WINDOW_WIDTH, 1, 0],
  ])(
    'swipe from %ipx to %ipx (cursor %i -> cursor %i) ',
    (from, to, cursorFrom, cursorTo) => {
      const { container } = render(<Component cursor={cursorFrom} />);
      const [carousel] = Array.from<any>(container.children);
      swipe(carousel, [from], [to]);
      expect(changeHandler).toHaveBeenCalledWith(cursorTo);
    }
  );
});

describe('check slides switching on swiping (Y axis)', () => {
  const changeHandler = jest.fn();
  const { Component } = buildTestCarousel({
    baseProps: {
      onCursorChange: changeHandler,
      y: true,
    },
  });

  test.each([
    [DEFAULT_WINDOW_HEIGHT, 0, 0, 1],
    [DEFAULT_WINDOW_HEIGHT, 0, 1, 2],
    [0, DEFAULT_WINDOW_HEIGHT, 2, 1],
    [0, DEFAULT_WINDOW_HEIGHT, 1, 0],
  ])(
    'swipe from %ipx to %ipx (cursor %i -> cursor %i) ',
    (from, to, cursorFrom, cursorTo) => {
      const { container } = render(<Component cursor={cursorFrom} />);
      const [carousel] = Array.from<any>(container.children);
      swipe(carousel, [0, from], [0, to]);
      expect(changeHandler).toHaveBeenCalledWith(cursorTo);
    }
  );
});

describe('check infinite switching', () => {
  const { Component } = buildTestCarousel({ baseProps: { infinite: true } });

  test.each([
    [100, 101],
    [101, 102],
  ])('move from cursor %i to cursor %i', (from, to) => {
    const { rerender, asFragment } = render(<Component cursor={from} />);
    const curFragment = asFragment();
    rerender(<Component cursor={to} />);
    expect(curFragment).toMatchDiffSnapshot(asFragment());
  });
});

test('check cursor shifts after adding or removing a slide in the infinite mode', () => {
  const changeHandler = jest.fn();
  const slidesCount = 3;
  const { Component } = buildTestCarousel({
    baseProps: {
      onCursorChange: changeHandler,
      defaultCursor: 100,
      infinite: true,
    },
    slidesCount,
  });

  const { rerender } = render(<Component />);
  rerender(<Component>{getTestSlides(slidesCount + 1)}</Component>);

  expect(changeHandler).toHaveBeenCalledWith(133);

  rerender(<Component>{getTestSlides(slidesCount)}</Component>);
  expect(changeHandler).toHaveBeenCalledWith(100);
});

test('check swipe with swipeThreshold', () => {
  const changeHandler = jest.fn();
  const thresholdPx = 30;

  const { Component } = buildTestCarousel({
    baseProps: {
      onCursorChange: changeHandler,
      swipeThreshold: `${thresholdPx}px`,
    },
  });

  const { container } = render(<Component />);
  const [carousel] = Array.from<any>(container.children);
  swipe(carousel, [DEFAULT_WINDOW_WIDTH], [DEFAULT_WINDOW_WIDTH - thresholdPx]);
  expect(changeHandler).toHaveBeenCalledWith(1);
});

describe('check displayAtOnce prop', () => {
  const visibleActorsChangeHandler = jest.fn();

  const { Component } = buildTestCarousel({
    baseProps: {
      displayAtOnce: [
        { value: 1 },
        { breakpoint: 800, value: 3 },
        { breakpoint: 400, value: 2 },
      ],
      onVisibleActorsChange: visibleActorsChangeHandler,
    },
  });

  test.each([
    [1000, 3],
    [500, 2],
    [200, 1],
  ])(
    'resizing window width to %ipx should display %i slide(s) at once',
    (width, visibleSlidesCount) => {
      resize(width, DEFAULT_WINDOW_HEIGHT);
      const { asFragment } = render(<Component />);
      expect(asFragment()).toMatchSnapshot();
      expect(visibleActorsChangeHandler).toHaveBeenCalledWith(
        visibleSlidesCount
      );
    }
  );
});

test('check trimEnd prop', () => {
  const changeHandler = jest.fn();

  const { Component } = buildTestCarousel({
    baseProps: {
      displayAtOnce: 3,
      onRangeChange: changeHandler,
    },
  });
  const { rerender } = render(<Component cursor={2} />);
  expect(changeHandler).toHaveBeenCalledWith(0, 2, 3);
  rerender(<Component cursor={2} trimEnd={true} />);
  expect(changeHandler).toHaveBeenCalledWith(0, 0, 3);
});

test('check autosize prop', () => {
  useOnResize.mockImplementation(
    getUseOnResizeMock([
      { width: 250, height: 0 },
      { width: 250, height: 0 },
      { width: 250, height: 20 },
      { width: 250, height: 30 },
      { width: 250, height: 10 },
      { width: 250, height: 40 },
    ])
  );

  const { Component } = buildTestCarousel({
    baseProps: {
      displayAtOnce: 2,
      autosize: true,
    },
  });
  const { rerender, asFragment } = render(<Component cursor={0} />);
  const curFragment = asFragment();

  useOnResize.mockImplementation(
    getUseOnResizeMock([
      { width: 250, height: 30 },
      { width: 250, height: 40 },
      { width: 250, height: 10 },
      { width: 250, height: 50 },
      { width: 250, height: 20 },
      { width: 250, height: 60 },
    ])
  );

  rerender(<Component cursor={2} />);
  expect(curFragment).toMatchDiffSnapshot(asFragment());
});
