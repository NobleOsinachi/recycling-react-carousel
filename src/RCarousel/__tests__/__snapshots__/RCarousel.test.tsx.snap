// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`check defaultCursor prop 1`] = `
<DocumentFragment>
  <div
    class="container touchActionNone w100 h100"
  >
    <div
      class="actor h100"
      style="width: 100%; transform: translate3d(-100%, 0%, 0);"
    >
      <div
        class="childWrapper w100 h100"
      >
        <div
          data-testid="slide2"
        >
          slide2
        </div>
      </div>
    </div>
    <div
      class="actor h100"
      style="width: 100%; transform: translate3d(0%, 0%, 0);"
    >
      <div
        class="childWrapper w100 h100"
      >
        <div
          data-testid="slide3"
        >
          slide3
        </div>
      </div>
    </div>
    <div
      class="actor h100"
      style="width: 100%; transform: translate3d(100%, 0%, 0);"
    />
  </div>
</DocumentFragment>
`;

exports[`check infinite switching move from cursor 100 to cursor 101 1`] = `
Snapshot Diff:
- First value
+ Second value

@@ -2,11 +2,11 @@
    <div
      class="container touchActionNone w100 h100"
    >
      <div
        class="actor h100"
-       style="width: 100%; transform: translate3d(-100%, 0%, 0);"
+       style="width: 100%; transform: translate3d(100%, 0%, 0);"
      >
        <div
          class="childWrapper w100 h100"
        >
          <div
@@ -16,11 +16,11 @@
          </div>
        </div>
      </div>
      <div
        class="actor h100"
-       style="width: 100%; transform: translate3d(0%, 0%, 0);"
+       style="width: 100%; transform: translate3d(-100%, 0%, 0);"
      >
        <div
          class="childWrapper w100 h100"
        >
          <div
@@ -30,11 +30,11 @@
          </div>
        </div>
      </div>
      <div
        class="actor h100"
-       style="width: 100%; transform: translate3d(100%, 0%, 0);"
+       style="width: 100%; transform: translate3d(0%, 0%, 0);"
      >
        <div
          class="childWrapper w100 h100"
        >
          <div
`;

exports[`check infinite switching move from cursor 101 to cursor 102 1`] = `
Snapshot Diff:
- First value
+ Second value

@@ -2,11 +2,11 @@
    <div
      class="container touchActionNone w100 h100"
    >
      <div
        class="actor h100"
-       style="width: 100%; transform: translate3d(-100%, 0%, 0);"
+       style="width: 100%; transform: translate3d(100%, 0%, 0);"
      >
        <div
          class="childWrapper w100 h100"
        >
          <div
@@ -16,11 +16,11 @@
          </div>
        </div>
      </div>
      <div
        class="actor h100"
-       style="width: 100%; transform: translate3d(0%, 0%, 0);"
+       style="width: 100%; transform: translate3d(-100%, 0%, 0);"
      >
        <div
          class="childWrapper w100 h100"
        >
          <div
@@ -30,11 +30,11 @@
          </div>
        </div>
      </div>
      <div
        class="actor h100"
-       style="width: 100%; transform: translate3d(100%, 0%, 0);"
+       style="width: 100%; transform: translate3d(0%, 0%, 0);"
      >
        <div
          class="childWrapper w100 h100"
        >
          <div
`;

exports[`check slides switching using cursor prop move from cursor 0 to cursor 1 1`] = `
Snapshot Diff:
- First value
+ Second value

@@ -2,15 +2,25 @@
    <div
      class="container touchActionNone w100 h100"
    >
      <div
        class="actor h100"
-       style="width: 100%; transform: translate3d(-100%, 0%, 0);"
-     />
+       style="width: 100%; transform: translate3d(100%, 0%, 0);"
+     >
+       <div
+         class="childWrapper w100 h100"
+       >
+         <div
+           data-testid="slide3"
+         >
+           slide3
+         </div>
+       </div>
+     </div>
      <div
        class="actor h100"
-       style="width: 100%; transform: translate3d(0%, 0%, 0);"
+       style="width: 100%; transform: translate3d(-100%, 0%, 0);"
      >
        <div
          class="childWrapper w100 h100"
        >
          <div
@@ -20,11 +30,11 @@
          </div>
        </div>
      </div>
      <div
        class="actor h100"
-       style="width: 100%; transform: translate3d(100%, 0%, 0);"
+       style="width: 100%; transform: translate3d(0%, 0%, 0);"
      >
        <div
          class="childWrapper w100 h100"
        >
          <div
`;

exports[`check slides switching using cursor prop move from cursor 1 to cursor 0 1`] = `
Snapshot Diff:
- First value
+ Second value

@@ -2,11 +2,11 @@
    <div
      class="container touchActionNone w100 h100"
    >
      <div
        class="actor h100"
-       style="width: 100%; transform: translate3d(-100%, 0%, 0);"
+       style="width: 100%; transform: translate3d(0%, 0%, 0);"
      >
        <div
          class="childWrapper w100 h100"
        >
          <div
@@ -16,11 +16,11 @@
          </div>
        </div>
      </div>
      <div
        class="actor h100"
-       style="width: 100%; transform: translate3d(0%, 0%, 0);"
+       style="width: 100%; transform: translate3d(100%, 0%, 0);"
      >
        <div
          class="childWrapper w100 h100"
        >
          <div
@@ -30,19 +30,9 @@
          </div>
        </div>
      </div>
      <div
        class="actor h100"
-       style="width: 100%; transform: translate3d(100%, 0%, 0);"
-     >
-       <div
-         class="childWrapper w100 h100"
-       >
-         <div
-           data-testid="slide3"
-         >
-           slide3
-         </div>
-       </div>
-     </div>
+       style="width: 100%; transform: translate3d(-100%, 0%, 0);"
+     />
    </div>
  </DocumentFragment>
`;

exports[`check slides switching using cursor prop move from cursor 1 to cursor 2 1`] = `
Snapshot Diff:
- First value
+ Second value

@@ -2,25 +2,15 @@
    <div
      class="container touchActionNone w100 h100"
    >
      <div
        class="actor h100"
-       style="width: 100%; transform: translate3d(-100%, 0%, 0);"
-     >
-       <div
-         class="childWrapper w100 h100"
-       >
-         <div
-           data-testid="slide1"
-         >
-           slide1
-         </div>
-       </div>
-     </div>
+       style="width: 100%; transform: translate3d(100%, 0%, 0);"
+     />
      <div
        class="actor h100"
-       style="width: 100%; transform: translate3d(0%, 0%, 0);"
+       style="width: 100%; transform: translate3d(-100%, 0%, 0);"
      >
        <div
          class="childWrapper w100 h100"
        >
          <div
@@ -30,11 +20,11 @@
          </div>
        </div>
      </div>
      <div
        class="actor h100"
-       style="width: 100%; transform: translate3d(100%, 0%, 0);"
+       style="width: 100%; transform: translate3d(0%, 0%, 0);"
      >
        <div
          class="childWrapper w100 h100"
        >
          <div
`;

exports[`check slides switching using cursor prop move from cursor 2 to cursor 1 1`] = `
Snapshot Diff:
- First value
+ Second value

@@ -2,11 +2,11 @@
    <div
      class="container touchActionNone w100 h100"
    >
      <div
        class="actor h100"
-       style="width: 100%; transform: translate3d(-100%, 0%, 0);"
+       style="width: 100%; transform: translate3d(0%, 0%, 0);"
      >
        <div
          class="childWrapper w100 h100"
        >
          <div
@@ -16,11 +16,11 @@
          </div>
        </div>
      </div>
      <div
        class="actor h100"
-       style="width: 100%; transform: translate3d(0%, 0%, 0);"
+       style="width: 100%; transform: translate3d(100%, 0%, 0);"
      >
        <div
          class="childWrapper w100 h100"
        >
          <div
@@ -30,9 +30,19 @@
          </div>
        </div>
      </div>
      <div
        class="actor h100"
-       style="width: 100%; transform: translate3d(100%, 0%, 0);"
-     />
+       style="width: 100%; transform: translate3d(-100%, 0%, 0);"
+     >
+       <div
+         class="childWrapper w100 h100"
+       >
+         <div
+           data-testid="slide1"
+         >
+           slide1
+         </div>
+       </div>
+     </div>
    </div>
  </DocumentFragment>
`;
